Tester/CodeCoverage/PhpParser.php

```diff
@@ -0,0 +1,183 @@
+<?php
+
+namespace Tester\CodeCoverage;
+
+
+/**
+ * Parses PHP source code and returns:
+ * - the start/end line information about functions, classes, interfaces, traits and their methods
+ * - the count of code lines
+ * - the count of commented code number
+ *
+ * @internal
+ */
+class PhpParser
+{
+	/**
+	 * @param  string  PHP code to analyze
+	 * @return \stdClass
+	 *
+	 * Returned structure is:
+	 *     stdClass {
+	 *         linesOfCode: int,
+	 *         linesOfComments: int,
+	 *         functions: [functionName => $functionInfo],
+	 *         classes: [className => $info],
+	 *         traits: [traitName => $info],
+	 *         interfaces: [interfaceName => $info],
+	 *     }
+	 *
+	 * where $functionInfo is:
+	 *     stdClass {
+	 *         start: int,
+	 *         end: int
+	 *     }
+	 *
+	 * and $info is:
+	 *     stdClass {
+	 *         start: int,
+	 *         end: int,
+	 *         methods: [methodName => $methodInfo]
+	 *     }
+	 *
+	 * where $methodInfo is:
+	 *     stdClass {
+	 *         start: int,
+	 *         end: int,
+	 *         visibility: public|protected|private
+	 *     }
+	 */
+	public function parse($code)
+	{
+		$tokens = @token_get_all($code);
+
+		$level = $classLevel = $functionLevel = NULL;
+		$namespace = '';
+		$line = 1;
+
+		$result = (object) array(
+			'linesOfCode' => max(1, substr_count($code, "\n")),
+			'linesOfComments' => 0,
+			'functions' => array(),
+			'classes' => array(),
+			'traits' => array(),
+			'interfaces' => array(),
+		);
+
+		$T_TRAIT = PHP_VERSION_ID < 50400 ? -1 : T_TRAIT;
+		while(list(, $token) = each($tokens)) {
+			if (is_array($token)) {
+				if (PHP_VERSION_ID < 50400 && $token[0] === T_STRING && strcasecmp($token[1], 'trait') === 0) {
+					$token[0] = $T_TRAIT;
+				}
+				$line = $token[2];
+			}
+
+			switch (is_array($token) ? $token[0] : $token) {
+				case T_NAMESPACE:
+					$namespace = ltrim(self::fetch($tokens, array(T_STRING, T_NS_SEPARATOR)) . '\\', '\\');
+					break;
+
+				case T_CLASS:
```

Comment:
@dg I guess yes, the `self::fetch(..., T_STRING)` condition follows. But I'll add it to tests.